## Major Project COMP8240

### Default Project - Group J

### Author: Sadia Tasnim (45760179)

###  fastText library exploration 

The objective of this project is to explore the facebook machine learning library called fastText which is doing a phenomenal job in word labelling and representation. This particular machine learning library can be applied to the purpose of text classification which is a necessity in case of sentiment analysis, business analysis (such as - product reviews, customer satisfaction percentage), as well as political situation analysis. In this project we will show our work in two steps - 
1. Replication of original work
2. Applying fasttext model in new dataset and comparison with the original work


The purpose is to show how the library can be used in supervised machine learning and even be tuned to be fast and more accurate.

Twitter data sources
1. http://cucis.ece.northwestern.edu/projects/Social/sentiment_data.html
2. http://help.sentiment140.com/for-students

## Loading the necessary libraries

In [1]:
import pandas as pd
import numpy as np

## Data Preprocessing

The Twitter dataset from the above mentioned first link have two separate files for twitter data and labels.The purpose is to combine those the two text files so that the labels for each tweet can be viewed in one line and the model setup in fasttext can be completed. 

I have taken the second dataset to enrich the volume and variety of the dataset as the first dataset had only 500 tweets. 

In [2]:
#This code will combine the tweets and the labels of the first data source
combine =[]

with open("tweets_data.txt") as xh:
    with open('tweets_label.txt') as yh:
        with open("tweetsShort.txt","w") as zh:
      #Read first file
          xlines = xh.readlines()
      #Read second file
          ylines = yh.readlines()
      #Combine content of both lists
      #combine = list(zip(ylines,xlines))
      #Write to third file
          for i in range(len(xlines)):
            line = "__label__"+ ylines[i].strip() + ' ' + xlines[i]
            zh.write(line)

In [3]:
#This code will read tweets from the second dataset
DATASET_COLUMNS=['target','ids','date','flag','user','text']
DATASET_ENCODING = "ISO-8859-1"
tweet_long_df = pd.read_csv('trainingandtestdata/tweetTraining.csv', encoding=DATASET_ENCODING, names=DATASET_COLUMNS)
tweet_long_df.sample(5)

target         ids                          date      flag  \
735969        0  2264907016  Sun Jun 21 05:16:35 PDT 2009  NO_QUERY   
588842        0  2216790764  Wed Jun 17 19:58:28 PDT 2009  NO_QUERY   
888346        4  1687276371  Sun May 03 08:15:31 PDT 2009  NO_QUERY   
1400927       4  2054470446  Sat Jun 06 07:25:25 PDT 2009  NO_QUERY   
845770        4  1564282388  Mon Apr 20 02:43:02 PDT 2009  NO_QUERY   

                  user                                               text  
735969         aimrere  headache now coz this prblem ! am not a rebel ...  
588842     krisshuntx3  @sicknastyemily aww thanks! But i think i have...  
888346   Laurabartleyy                           gettin ready for work!!   
1400927         lgreps  just enjoyed the best 2 hour work out ever. we...  
845770       parasight  What a hilarious music video.  Mickey 3D - Mat...

In [4]:
# Creating a new dataframe with only the tweets 
tweet_long_df_new = tweet_long_df.filter(['text'], axis=1)

In [5]:
# printing the dataframe
tweet_long_df_new.head()

text
0  @switchfoot http://twitpic.com/2y1zl - Awww, t...
1  is upset that he can't update his Facebook by ...
2  @Kenichan I dived many times for the ball. Man...
3    my whole body feels itchy and like its on fire 
4  @nationwideclass no, it's not behaving at all....

In [6]:
# Saving the tweets (every tweet in a new line) in a text file
tweet_long_df_new.to_csv('longTweet.txt', sep='\n', index=False)

In [7]:
# Creating a new dataframe with only the labels
tweet_long_df_new_label = tweet_long_df.filter(['target'], axis=1)

According to the dataset description, the polarity of the tweets are saved in numeric form. 0 = negative, 2 = neutral, 4 = positive. However, the dataset from link 1 saves the polarity as such- N = negative, O = neutral, P = positive. To make the datasets of the same format, I am changing the polarity of the second dataset according to the first dataset format.

In [8]:
tweet_long_df_new_label.loc[(tweet_long_df_new_label.target == 0), 'target'] = 'N'

In [9]:
tweet_long_df_new_label.loc[(tweet_long_df_new_label.target == 2), 'target'] = 'O'

In [10]:
tweet_long_df_new_label.loc[(tweet_long_df_new_label.target == 4), 'target'] = 'P'

In [11]:
tweet_long_df_new_label

target
0            N
1            N
2            N
3            N
4            N
...        ...
1599995      P
1599996      P
1599997      P
1599998      P
1599999      P

[1600000 rows x 1 columns]

In [12]:
# Saving the label tweets (every tweet in a new line) in a text file
tweet_long_df_new_label.to_csv('twitterLabelBig.txt', sep='\n', index=False)

Next step is to merge the tweets from the different datasets

In [13]:
# Python program to merge two files 
dataTweet = dataTweet2 = "" 
  
# Reading data from first file 
with open('tweets_data.txt', errors='ignore') as fp: 
    dataTweet = fp.read() 
    with open('longTweet.txt', errors='ignore') as fp: 
        dataTweet2 = fp.read() 
# Merging two files into one another file 
dataTweet += "\n"
dataTweet += dataTweet2 
with open ('totalTweet.txt', 'w') as fp: 
    fp.write(dataTweet)

Now merging the labels from the two datasets

In [15]:
# Python program to merge two files 
dataTweet = dataTweet2 = "" 
  
# Reading data from first file 
with open('tweets_label.txt', errors='ignore') as fp: 
    dataTweet = fp.read() 
    with open('twitterLabelBig.txt', errors='ignore') as fp: 
        dataTweet2 = fp.read() 
# Merging two files into one another file 
dataTweet += "\n"
dataTweet += dataTweet2 
with open ('totalLabel.txt', 'w') as fp: 
    fp.write(dataTweet)

The following step combines the tweets with their polarity value. The prefix "__label__" is necessary for the fasttext format

In [16]:
combine =[]

with open("totalTweet.txt") as xh:
    with open('totalLabel.txt') as yh:
        with open("totalLabelledTweet.txt","w") as zh:
      #Read first file
          xlines = xh.readlines()
      #Read second file
          ylines = yh.readlines()
      #Combine content of both lists
      #combine = list(zip(ylines,xlines))
      #Write to third file
          for i in range(len(xlines)):
            line = "__label__"+ ylines[i].strip() + ' ' + xlines[i]
            zh.write(line)

The command line code "wc totalLabelledTweet.txt" will give us the result of the word counts, line counts information which is in my case is 1600500 (line count or in this case number of tweets)

The next step is to divide the datasets into train and test sets. I have created tweet.train and tweet.valid sets for training and testing purpose respectively.

In [17]:
# importing fasttext library
import fasttext

train_supervised() will mostly be used for retruning a model object and calling test and predict on that object. This is the same as learning the text classifier.

In [18]:
model = fasttext.train_supervised(input="tweet.train")

The input argument indicates the file containing the training examples. We can now use the model variable to access information on the trained model.

We can also call save_model to save it as a file and load it later with load_model function.
<br>source: https://fasttext.cc/docs/en/supervised-tutorial.html 

In [19]:
model.save_model("model_tweet.bin")

For testing the classifier, model.predict() is used.

In [20]:
model.predict("Because I'm happy")

(('__label__P',), array([0.98308438]))

As happy is a positive word, it is labelled as P

In [21]:
model.predict("Which baking dish is best to bake a banana bread ?")

(('__label__P',), array([0.96914047]))

The word "best" is a positive word

In [23]:
model.predict("I am sad")

(('__label__N',), array([1.00000954]))

Smilarly "sad" is a word discribing negative emotion 

Next step is to test the model's quality on the test set

In [24]:
model.test("tweet.valid")

(480500, 0.6430884495317378, 0.6430884495317378)

The output are the number of samples (here 480500), the precision at one (0.643) and the recall at one (0.643).

In [25]:
model.test("tweet.valid", k=5)

(480500, 0.3333333333333333, 1.0)

The precision at 5 is 0.333 and recall is 1

As we know that, precision discloses the number of correct labels among the labels predicted by fastText and the recall score is the number of successfully predicted labels among all the real ones.

In [26]:
model.predict("Because I'm happy", k=5)

(('__label__P', '__label__N', '__label__O'),
 array([9.83084381e-01, 1.69356670e-02, 1.00003181e-05]))

## Model Betterment

A crude normalization can be obtained using the command line tools such as sed and tr and by using that, I removed the uppercase and punctuation marks.

After preprocessing, I tried to train the new model on that data

In [33]:
model_preprocssed = fasttext.train_supervised(input="tweetL.train")

In [34]:
model_preprocssed.test("tweetL.valid")

(480500, 0.12849947970863684, 0.12849947970863684)

###  Model betterment through increased epoch

In [35]:
model_hyperparameter_tuned_epochs = fasttext.train_supervised(input="tweetL.train", epoch=50)

In [37]:
model_hyperparameter_tuned_epochs.test("tweetL.valid")

(480500, 0.7870697190426639, 0.7870697190426639)

After increasing the number of epochs the model became quite strong. By setting up the epoch option, I increased the number of times each example is seen from default value 5 to 50. 

### Setting up larger learning rate

In [38]:
model_hyperparameter_tuned_lr = fasttext.train_supervised(input="tweetL.train", lr=1.0)

In [39]:
model_hyperparameter_tuned_lr.test("tweetL.valid")

(480500, 0.6340457856399584, 0.6340457856399584)

precision and recall score gets a bit down from the scores of increased epoch ( around 19%)

In [40]:
model_hyperparameter_tuned_lr1 = fasttext.train_supervised(input="tweetL.train", lr=0.5)

In [41]:
model_hyperparameter_tuned_lr1.test("tweetL.valid")

(480500, 0.43113423517169613, 0.43113423517169613)

As we know the range of lr is from 0.1 to 1.0. By applying lr=0.5, I could see that the scores decreased

### Applying learing rate and increased epoch

In [47]:
model_hyperparameter_tuned = fasttext.train_supervised(input="tweetL.train", lr=1.0, epoch=25)

In [48]:
model_hyperparameter_tuned.test("tweetL.valid")

(480500, 0.5827825182101977, 0.5827825182101977)

## Word n-grams

In case of sentiment analysis, using bigrams can be beneficial as word order can effect the accuracy of sentiment analysis. For example, In the sentence- "I am not good at accounting." has the word "good" in it, it might be declared as positive but, using bigrams can be useful here as the word order "not good" will be taken into account and deemed as negative. 

In [45]:
model_bigram = fasttext.train_supervised(input="tweetL.train", lr=1.0, epoch=25, wordNgrams=2)

In [46]:
model_bigram.test("tweetL.valid")

(480500, 0.6259791883454735, 0.6259791883454735)

## Hierarchical softmax

A potential solution for faster training is to use hierarchical softmax instead of the regular softmax. This can be utilised with the option -loss hs :

In [49]:
model_hisoftmax = fasttext.train_supervised(input="tweetL.train", lr=1.0, epoch=25, wordNgrams=2, bucket=200000, dim=50, loss='hs') 

In [51]:
model_hisoftmax.test("tweetL.valid")

(480500, 0.802834547346514, 0.802834547346514)

The increase rate in precision from bigram to Hierarchical softmax is 22.06% 

The hierarchical softmax is a loss function that approximates the softmax with a much faster computation.

The idea is to build a binary tree whose leaves correspond to the labels. Each intermediate node has a binary decision activation (e.g. sigmoid) that is trained, and predicts if we should go to the left or to the right. The probability of the output unit is then given by the product of the probabilities of intermediate nodes along the path from the root to the output unit leave.

For a detailed explanation, you can have a look on this video.

In fastText, we use a Huffman tree, so that the lookup time is faster for more frequent outputs and thus the average lookup time for the output is optimal.
<br>
Data Source: https://fasttext.cc/docs/en/supervised-tutorial.html

## Multi-label classification

This step is beneficial when we want to assign multiple labels to the document. When we use independent binary classifiers for each label, it creates a convenient environment to handle multple labels. 

In [52]:
model_multiClass= fasttext.train_supervised(input="tweetL.train", lr=0.5, epoch=25, wordNgrams=2, bucket=200000, dim=50, loss='ova')

In [53]:
model_multiClass.predict("because I'm not happy", k=-1, threshold = 0.5)

(('__label__n',), array([1.00001001]))

In [54]:
model_multiClass.test("tweetL.valid", k=-1)

(480500, 0.3333333333333333, 1.0)

Now it shows that we are getting recall score 1.0 